packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/service/SharedPreferencesNotificationsStore.java

```diff
@@ -0,0 +1,189 @@
+   *
+   * @param identifier Identifier of the notification.
+   * @return Notification information: request and trigger.
+   * @throws JSONException          Thrown if notification request could not have been interpreted as a JSON object.
+   * @throws IOException            Thrown if there is an error when fetching trigger from the storage.
+   * @throws ClassNotFoundException Thrown if there is an error when interpreting trigger fetched from the storage.
+   */
+  public Pair<JSONObject, SchedulableNotificationTrigger> getNotification(String identifier) throws JSONException, IOException, ClassNotFoundException {
+    JSONObject request = new JSONObject(mSharedPreferences.getString(preferencesRequestKey(identifier), null));
+    SchedulableNotificationTrigger trigger = deserializeTrigger(mSharedPreferences.getString(preferencesTriggerKey(identifier), null));
+    return new Pair<>(request, trigger);
+  }
+
+  /**
+   * Fetches all scheduled notifications, ignoring invalid ones.
+   * <p>
+   * Goes through all the {@link SharedPreferences} entries, interpreting only the ones conforming
+   * to the expected format.
+   *
+   * @return Map with identifiers as keys and notification info as values
+   */
+  public Map<String, Pair<JSONObject, SchedulableNotificationTrigger>> getAllNotifications() {
+    Map<String, Pair<JSONObject, SchedulableNotificationTrigger>> allNotifications = new HashMap<>();
+
+    for (Map.Entry<String, ?> entry : mSharedPreferences.getAll().entrySet()) {
+      String key = entry.getKey();
+      try {
+        String identifier;
+        if (key.startsWith(REQUEST_KEY_PREFIX)) {
+          identifier = key.substring(REQUEST_KEY_PREFIX.length());
+        } else if (key.startsWith(TRIGGER_KEY_PREFIX)) {
+          identifier = key.substring(TRIGGER_KEY_PREFIX.length());
+        } else {
+          continue;
+        }
+
+        JSONObject request = null;
+        SchedulableNotificationTrigger trigger = null;
+        Pair<JSONObject, SchedulableNotificationTrigger> scheduledNotification = allNotifications.get(identifier);
+        if (scheduledNotification != null) {
+          request = scheduledNotification.first;
+          trigger = scheduledNotification.second;
+        }
+
+        if (key.startsWith(REQUEST_KEY_PREFIX)) {
+          request = new JSONObject((String) entry.getValue());
+        } else if (key.startsWith(TRIGGER_KEY_PREFIX)) {
+          trigger = deserializeTrigger((String) entry.getValue());
+        }
+        allNotifications.put(identifier, new Pair<>(request, trigger));
+      } catch (JSONException | ClassNotFoundException | IOException e) {
+        // do nothing
+      }
+    }
+
+    // Remove invalid notifications
+    for (String key : allNotifications.keySet()) {
+      Pair<JSONObject, SchedulableNotificationTrigger> pair = allNotifications.get(key);
+      if (pair == null || pair.first == null || pair.second == null) {
+        allNotifications.remove(key);
```

Comment:
maybe worth deleting from persistent storage at this point, too?

Reply:
I was wondering about that too, at first I decided not to:
- not to modify the store in `getAllNotifications` (a GET)
- it may happen that `pair.first == null || pair.second == null` if eg. a developer would first save the notification with a custom trigger, then upgrade to a version without a custom trigger, then, wanting to fix his problem, upgrade again

I guess in that case one could as well just reschedule the notifications. Fixed!